### ISSUE TITLE
Isolating one Homestead VM of a duplex pair from the signalling network causes call failures for 20 seconds.

### ISSUE THREAD

**Issue by plfrancis** [Metaswitch/homestead#268]
While running a 25 CPS traffic load, isolated a Homestead VM from the signalling network. This resulted in 500 call failures over the next 20 seconds. The cause was Sprout timing out waiting for registration state and IFCs from Homestead. Homestead in turn was waiting on Cassandra. For the first number of failures Homestead simply reported the following after 14 seconds:

```
Failed to get registration information from the cache - reject the request
```

After this, Homestead started reporting the following after 18 seconds:

```
Insufficient nodes to read Cassandra at consistency level LOCAL_QUORUM
Insufficient nodes to read Cassandra at consistency level QUORUM
```

Subsequent errors were of the previous variety with the time between request and Cassandra error response decreasing to 0 over the next 18 seconds. At that point calls were once again processed normally. Diagnostics will be uploaded.

**Comment by mirw** [comment]
It looks as though `LOCAL_QUORUM` and `QUORUM` reads can be very slow in this window, and this means that we hold up the `ONE` read for a long time (until the original HTTP request has long timed out).

I suspect this has been introduced by https://github.com/Metaswitch/cpp-common/commit/bf130d194058c1ed583131db80b48bfff9352d9f.

**Comment by eleanor-merry** [comment]
Hi @plfrancis - passing this back to you to determine if taking the fix is worth while.

**Comment by plfrancis** [comment]
Hi @eleanor-merry - with the fix it looks like Homestead no longer waits for Cassandra for long periods. But, calls still fail for 20 seconds after a Homestead VM is isolated from the network. The new issue seems to be that Sprout and Homestead are both using a 1/2 second timer and Sprout times out first before Homestead can return its 504 response (just milliseconds later). I've uploaded a SAS trace showing the new scenario.

**Comment by plfrancis** [comment]
For some reason I was thinking if Sprout received the response from Homestead before timing out it would behaved differently. But, for this scenario that is probably not the case. The real question I guess is why Homestead timed out reading the cache when one Cassandra instance is still available.

**Comment by AlexHockey** [comment]
A couple of questions for you @plfrancis:

* Is your deployment single-site or GR?
* Can you paste your current cassandra.yaml into the issue (there is one in the diags bundle you uploaded in October but I want to make sure I'm looking at the latest version).

**Comment by AlexHockey** [comment]
Can you also let me know the output of running this command on your homestead nodes: `target_latency_us=100000; . /etc/clearwater/config; echo -n $target_latency_us`?

**Comment by plfrancis** [comment]
@AlexHockey, the deployment is single-site. Here is a copy of cassandra.yaml:

```
# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
#
# WARNING - THIS FILE IS GENERATED BY ETCD AND SHOULD NOT BE EDITED DIRECTLY
#
# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
authenticator: AllowAllAuthenticator
authorizer: AllowAllAuthorizer
auto_snapshot: true
batch_size_warn_threshold_in_kb: 5
batchlog_replay_throttle_in_kb: 1024
cas_contention_timeout_in_ms: 1000
client_encryption_options: {enabled: false, keystore: conf/.keystore, keystore_password: cassandra}
cluster_name: Test Cluster
column_index_size_in_kb: 64
commit_failure_policy: stop
commitlog_directory: /var/lib/cassandra/commitlog
commitlog_segment_size_in_mb: 32
commitlog_sync: periodic
commitlog_sync_period_in_ms: 10000
compaction_preheat_key_cache: true
compaction_throughput_mb_per_sec: 16
concurrent_reads: 32
concurrent_writes: 32
cross_node_timeout: false
data_file_directories: [/var/lib/cassandra/data]
disk_failure_policy: stop
dynamic_snitch_badness_threshold: 0.1
dynamic_snitch_reset_interval_in_ms: 600000
dynamic_snitch_update_interval_in_ms: 100
endpoint_snitch: GossipingPropertyFileSnitch
hinted_handoff_enabled: true
hinted_handoff_throttle_in_kb: 1024
in_memory_compaction_limit_in_mb: 64
incremental_backups: false
inter_dc_tcp_nodelay: false
internode_compression: all
key_cache_save_period: 14400
key_cache_size_in_mb: null
listen_address: 192.168.140.196
max_hint_window_in_ms: 10800000
max_hints_delivery_threads: 2
memtable_flush_queue_size: 4
multithreaded_compaction: false
native_transport_port: 9042
num_tokens: 256
partitioner: org.apache.cassandra.dht.Murmur3Partitioner
permissions_validity_in_ms: 2000
preheat_kernel_page_cache: false
range_request_timeout_in_ms: 10000
read_request_timeout_in_ms: 400
request_scheduler: org.apache.cassandra.scheduler.NoScheduler
request_timeout_in_ms: 10000
row_cache_save_period: 0
row_cache_size_in_mb: 0
rpc_address: localhost
rpc_keepalive: true
rpc_port: 9160
rpc_server_type: sync
saved_caches_directory: /var/lib/cassandra/saved_caches
seed_provider:
- class_name: org.apache.cassandra.locator.SimpleSeedProvider
  parameters:
  - {seeds: 192.168.140.196}
server_encryption_options: {internode_encryption: none, keystore: conf/.keystore,
  keystore_password: cassandra, truststore: conf/.truststore, truststore_password: cassandra}
snapshot_before_compaction: false
ssl_storage_port: 7001
start_native_transport: true
start_rpc: true
storage_port: 7000
thrift_framed_transport_size_in_mb: 15
tombstone_failure_threshold: 100000
tombstone_warn_threshold: 1000
trickle_fsync: false
trickle_fsync_interval_in_kb: 10240
truncate_request_timeout_in_ms: 60000
write_request_timeout_in_ms: 2000
```
The one on the other Homestead is exactly the same except for 'listen_address' which contains its signaling IP. The output of your command on both Homsteads is '100000'.

**Comment by AlexHockey** [comment]
Ok so the way this is supposed to work is:

* Cassandra is configured to time out all read operations after 400ms (`read_request_timeout_in_ms`). 
* Homestead starts by doing a TWO read (i.e. one that requires at least two replicas to respond before it can succeed). In a non-GR system this will fail if one homestead is isolated. At worst it will take 400ms to fail. 
* Homestead catches this failure and retries a (less consistent) ONE read, which should succeed (unless both replicas are down which is a double failure). 

The idea is that all of this can happen in less than 500ms - so before sprout times out the HTTP request. 

Homestead has a bug which means it doesn't always spot when the TWO read times out, and just fails the request instead of retrying a ONE read. However this doesn't explain why Phil was seeing homestead timeout the request after *500*ms - I would have expected it to timeout after 400ms. Maybe cassandra wasn't restarted and was running with old config? I'm going to fix the bug I know about. When Phil does a restest we can get more information about what config cassandra is actually running with.

**Comment by AlexHockey** [comment]
I've made a fix that means we should spot when a timeout occurs. @plfrancis please can you retest in release-87? When you retest please can you confirm that:

* `read_request_timeout_in_ms` is still set to 400ms. 
* Cassandra has been restarted since `cassandra.yaml` was last updated. 

I want to rule out the possibility that cassandra is using out-of-date config.

**Comment by JimBunch** [comment]
For some reason I cannot mark this issue as critical. It should be marked as critical because it is call processing service affecting.

**Comment by bossmc** [comment]
@JimBunch Looks like we'd forgotten to add a `critical` label to homestead.  I've added one an tagged this issue.

**Comment by tcaterisano** [comment]
Retested this and see call failures for 20 seconds.  SAS reports Cassandra operation failed due to lack of available nodes.  Observed a few hundred call failures at 20 CPS - 90 seconds hold time.

Two Cassandra settings eliminated all errors when one homestead VM in a two node cluster was stopped. Not certain if these could cause other issues but does correct this issue. 

phi_convict_threshold: 5
read_request_timeout_in_ms: 100

Please advise if you know of any problems these cassandra settings might cause.

**Comment by mirw** [comment]
I'm nervous about this change because we don't understand why @AlexHockey's fix above doesn't work.

@tcaterisano, please can you answer Alex's questions above...

> When you retest please can you confirm that:
>
> *   read_request_timeout_in_ms is still set to 400ms.
> *   Cassandra has been restarted since cassandra.yaml was last updated.

...and then assign back to him?

**Comment by mirw** [comment]
(If we do decide to go the route described above, we need to make sure we update all our documentation and config generation scripts to effect this change.)

**Comment by tcaterisano** [comment]
The answer to Alex's question is yes, Cassandra was restarted with the expected settings.

I've removed the setting - phi_convict_threshold: 5 and restarted Cassandra with only read_request_timeout_in_ms: 100.  I retested this saw no failures the first couple of tries. On the third try, I noticed 21 404 errors.  These all occurred in the same second so it is still possible to have some call failures but nothing like to 20 outage we saw with  read_request_timeout_in_ms: 400.

I've uploaded sas traces as requested for a call that worked and one that failed immediately after isolating homestead from the signaling network.

**Comment by AlexHockey** [comment]
Unclaiming as I don;t have time to work on this immediately.

**Comment by AlexHockey** [comment]
On the successful call I get these SAS logs from homestead. 

```
2016-06-07 16:30:04.147 homestead@homestead-2.vm7.richlab.datcon.co.uk Querying cache for registration data
2016-06-07 16:30:04.147 homestead@homestead-2.vm7.richlab.datcon.co.uk Cassandra operation failed due to lack of available nodes.
2016-06-07 16:30:04.147 homestead@homestead-2.vm7.richlab.datcon.co.uk Found registration data in the cache
...
2016-06-07 16:30:04.170 homestead@homestead-2.vm7.richlab.datcon.co.uk Querying cache for registration data
2016-06-07 16:30:04.171 homestead@homestead-2.vm7.richlab.datcon.co.uk Cassandra operation failed due to lack of available nodes.
2016-06-07 16:30:04.171 homestead@homestead-2.vm7.richlab.datcon.co.uk Found registration data in the cache
...
2016-06-07 16:30:04.183 homestead@homestead-2.vm7.richlab.datcon.co.uk Querying cache for registration data
2016-06-07 16:30:04.184 homestead@homestead-2.vm7.richlab.datcon.co.uk Cassandra operation failed due to lack of available nodes.
2016-06-07 16:30:04.185 homestead@homestead-2.vm7.richlab.datcon.co.uk Found registration data in the cache
```

i.e. the requests take about 1ms to timeout. On the failed call I get:

```
2016-06-08 16:30:07.500 homestead@homestead-2.vm7.richlab.datcon.co.uk Querying cache for registration data
2016-06-08 16:30:07.905 homestead@homestead-2.vm7.richlab.datcon.co.uk Cassandra operation failed due to lack of available nodes.
2016-06-08 16:30:07.906 homestead@homestead-2.vm7.richlab.datcon.co.uk Found registration data in the cache

2016-06-08 16:30:07.920 homestead@homestead-2.vm7.richlab.datcon.co.uk Querying cache for registration data
2016-06-08 16:30:08.501 homestead@homestead-2.vm7.richlab.datcon.co.uk Cassandra operation failed due to lack of available nodes.
2016-06-08 16:30:08.501 homestead@homestead-2.vm7.richlab.datcon.co.uk Found registration data in the cache
```

Once request takes ~400ms to timeout and one takes ~520. Sprout is only willing to wait for 500ms for a response so the last request causes it to timeout. 

It looks like Cassandra is just not honoring the configured `read_request_timeout_in_ms` parameter. Surely not?

**Comment by AlexHockey** [comment]
I think we need to understand where and why the delay is being introduced. 

@tcaterisano please can you set the `read_request_timeout_in_ms` back to 400ms and reproduce the problem while capturing traffic on cassandra's ports (`tcpdump -s0 -w <outputfile> -i any port 7000 or port 9160`). Please could you also grab a diags bundle after you've finished your test (as this will contain historical CPU stats which I want to look at).

**Comment by tcaterisano** [comment]
requested tcpdumps and diags uploaded

**Comment by tcaterisano** [comment]
Passing back to Alex for analysis

**Comment by rkd-msw** [comment]
@tcaterisano, it looks like the .pcap files are empty. Maybe you were taking the trace in the management namespace?

Assigning back to you to re-capture these diagnostics, but let me know if I can help.

**Comment by sathiyan-sivathas** [comment]
Next action is to collect tcpdump of this issue, so we're blocked until that can happen.

**Comment by rkd-msw** [comment]
I can retest this, I'm running stress at the moment.

**Comment by rkd-msw** [comment]
OK, I've attached a (non-empty!) pcap trace to \\ad\csd\bigfiles\clearwater\issue-attachments\homestead\268. The first call failed at 18:41:51.754926 (which was when I disconnected the signaling network adapter on homestead-1) and the last call failed at 18:42:09.029578 - at this point, calls started succeeding again without my reattaching anything.

I've also attached SAS trace of the first failed call.

This is on version `10.0.0-20161116-~255.00.0-161116.174458`. I have checked that config option:

```
clearwater@stress-homestead-1:~$ grep read_r /etc/cassandra/cassandra.yaml
read_request_timeout_in_ms: 400
```

and this was a system newly deployed on CCv10, so there's no risk of it running with old config.

I was using a much lower rate of 2.3 calls/second, so that the PCAP trace wouldn't be huge - so I don't think historical CPU stats will be very interesting but I've attached a diags bundle anyway.

**Comment by rkd-msw** [comment]
On reflection, taking network capture from homestead-1 (the node I disconnected from the network) probably isn't very informative. I've rerun the test - iss268_hs2.pcap contains packet capture from homestead-2 while homestead-1 was disconnected from the network. Calls failed between 19:05:19.391188 and 19:05:36.225824.

homestead-1 is 10.225.166.46, homestead-2 is 10.225.166.47.

**Comment by rkd-msw** [comment]
Three new SAS traces at `\\ad\csd\bigfiles\clearwater\issue-attachments\homestead\268`, as @eleanor-merry requested:

* one at the start of the 20s window, where reads fail with THRIFT_EAGAIN
* one at the end of the 20s window, where reads work but are so slow that Sprout times them out
* one after the 20s window, where it works

**Comment by rkd-msw** [comment]
I'm seeing the same symptoms when I power off one GR site (containing 2 out of my 4 Cassandras) and power it back on:

* site powers off between 14:54:11 and 14:54:26
* site powers back on at 14:56:26
* it takes a little longer for the processes to start up, until 14:58 or so, e.g. this log from Cassandra: `14:58:21,623 Gossiper.java:1027 - Node /10.225.166.169 has restarted, now UP`

* I see call failures with a 404 Not Found at the following times:

14:54:19.905073
14:54:21.000954
14:54:22.109083
14:54:23.217351
14:54:26.540897
14:54:28.752245
14:54:29.860254
14:54:30.969009
14:54:32.077252
14:54:33.185007
14:54:34.292234
14:54:35.400241
14:54:37.616777
14:54:38.721023
14:54:42.546560
14:55:16.889080
14:55:37.931016
14:56:12.270969
14:56:33.327720
14:57:07.655247
14:57:28.709201
14:58:03.038365

**Comment by rkd-msw** [comment]
(The 14:57:07.655247 example is attached as "GR failure trace.sas")

**Comment by eleanor-merry** [comment]
We've agreed to defer this until a later release

**Comment by rkd-msw** [comment]
"It looks like Cassandra is just not honoring the configured read_request_timeout_in_ms parameter. Surely not?"

Is it possible that we actually need to set `range_request_timeout_in_ms` here? Looking at cpp-common we're actually using a SliceRange to get multiple columns.
